### ISSUE TITLE
Implement `revalidatePath` functionality for the App Router

### ISSUE THREAD

**Issue by better-salmon** [caching-tools/next-shared-cache#382]

**Comment by duanv** [comment]
@better-salmon I've been trying to find what the limitation/reason is for why `revalidatePath` is not available for the App Router. Could you perhaps give us a TLDR of the limitation or am I misreading the docs and it's a feature that is possible just not implemented yet?

**Comment by better-salmon** [comment]
@duanv hello!

TLDR: this feature is possible but challenging to implement efficiently.

More specifically, the `revalidatePath` function is just a smart alias for the `revalidateTag`. Internally, it triggers the `revalidateTag` method of the `CacheHandler` object with a specific tag (soft tag). The soft tag exists in the tags list of `PAGE` kind cache entries but not in the `FETCH` kind. It means that after you trigger the `revalidatePath`, `CacheHandler` deletes `PAGE` from the cache, but the `FETCH` cache remains. As a result, the page with such a path will be re-rendered with old data and saved to the cache instead of being revalidated.

To implement it efficiently, we need to know the soft tags when we save `FETCH` kind entries in the cache. As far as I know, this is only possible with some hacks or changes in the Next.js code.

In the earlier versions, the revalidation process functioned differently. The `revalidatePath` and `revalidateTag` only marked some tags as revalidated but didn't delete them from the cache. Upon retrieving the data, `CacheHandler` checked if it was revalidated.

More context from the Next.js PR: https://github.com/vercel/next.js/pull/53321

**Comment by joshudev** [comment]
Thanks for the above info. We're hitting this problem right now.

Is there a workaround we can apply for now? Such as tagging each page with the path?

Also, is unstable_cache supported?

**Comment by franco14lorenzo** [comment]
I have the same problem, when I call revalidatePath('/', 'layout'), for example, my cache is not invalidated and the site continues to show the old version. If you have a solution, can you share it?

**Comment by duanv** [comment]
Thank you @better-salmon I appreciate the breakdown 👏

**Comment by better-salmon** [comment]
Hey @franco14lorenzo ! For now, you can only use revalidateTag(tag), where the tag is the tag from your fetch that is used on this page.

I'm working on a proper solution right now.

**Comment by better-salmon** [comment]
@joshudev hello! I've created an issue for `unstable_cache`. See #391

**Comment by szaveri** [comment]
@better-salmon I was thinking of a possible solution for the `revalidatePath` vs. `revalidateTag` issue and thought we could potentially utilize redis's `hash-tags` to group the keys and delete them by slot on revalidate. https://redis.io/docs/reference/cluster-spec/#hash-tags

Is it possible to get the associated path of a `fetch` call and create a key like

```
<path | tag>.<build number>.<some uuid?>{<path>.<build number>}
```

then when doing a `revalidatePath` we call 

```
await client.sendCommand(['CLUSTER', DELSLOTS', '<path>.<build number>']);
```

**EDIT**: Alternatively, maybe `HSET`, `HGETALL` and `UNLINK` might be better?

**Comment by ezeparziale** [comment]
For now, on all the pages where I have to use revalidatePath, I set them as `export const dynamic = 'force-dynamic'`
Until this issue is resolved. I'm not sure if anyone has another temporary solution.

**Comment by ezeparziale** [comment]
@better-salmon I ran some tests and was able to delete the pages and routes stored in Redis.

I'm not sure if it can be integrated into the CacheHandler process or if there are other cases to consider.

```typescript
// test-revalidatePath.ts

import { createClient } from "redis";

const client = createClient();

client.on("error", (err) => console.log("Redis Client Error", err));

const NEXT_CACHE_IMPLICIT_TAG_ID = "_N_T_";

async function revalidatePath(originalPath: string, type?: "layout" | "page") {
  await client.connect();

  const keys = await client.keys("*");

  let normalizedPath = `${NEXT_CACHE_IMPLICIT_TAG_ID}${originalPath}`;

  if (type) {
    normalizedPath += `${normalizedPath.endsWith("/") ? "" : "/"}${type}`;
  }

  for (const key of keys) {
    const typeKey = await client.TYPE(key);

    if (typeKey === "ReJSON-RL") {
      const value = await client.json.get(key);

      // @ts-ignore
      if (value?.value) {
        // @ts-ignore
        const kind = value["value"]["kind"];

        if (kind === "PAGE" || kind === "ROUTE") {
          // @ts-ignore
          const xNextCacheTags = value["value"]["headers"]["x-next-cache-tags"];

          if (xNextCacheTags.includes(normalizedPath)) {
            await client.del(key);
            console.log(`Key ${key} deleted`);
          }
        }
      }
    }
  }

  await client.disconnect();
}

revalidatePath("/", "layout");
```

**Comment by szaveri** [comment]
Similar concept to @ezeparziale 

I went with a slightly different solution than I was imaging earlier. There are basically two main types of cache we store in redis

1. Full Route Cache - `PAGE`
2. Data Cache - `FETCH`

There are a few others types: `IMAGE`, `ROUTE`, and `REDIRECT`, but I haven't fully tested those yet, though the only one I'll experiment with is `ROUTE`

My thoughts were that a `PAGE` cache should automatically be cleared if any of the associated `FETCH` are cleared, not the other away around. For example if you clear a `FETCH` tag `foo` then all `PAGE` cache associated with that tag should also be removed as well to get the desired affect. The page will update based on the new data validate, but all other `FETCH` will still be cached for optimized loading.

Here's the solution I came up with:

```javascript
// cache-handler.js
const { createClient } = require('redis');

const NEXT_PREFIX = '_N_T_';
const DEFAULT_TTL = 300;

const client = createClient();
client.on('error', err => console.log('Redis Client Error', err));
client.connect();
 
module.exports = class CacheHandler {
  constructor(options) {
    this.options = options
  }
 
  async get(key) {
    // This could be stored anywhere, like durable storage
    let value = await client.get(key);
    if (value) {
      value = JSON.parse(value);
    }

    return value;
  }
 
  async set(key, data, ctx) {
    /**
     * Special handler for PAGE caching to ensure we 
     * invalidate PAGE caching when clearing FETCH cache.
     * This is to ensure if we clear any FETCH within a page
     * it will also invalidate any page associted with that tag
     * 
     * Revalidate is based on lowest revalidate value on page
     * https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#revalidating-data
     */
    if (data?.kind === 'PAGE' && data?.status === 200) {
      ctx.tags = data.headers['x-next-cache-tags'].split(',').filter(tag => !tag.startsWith(NEXT_PREFIX) && tag !== '');
    }
    
    /**
     * We're creating a mapping of all the keys
     * associted with a give FETCH tag
     */
    ctx.tags.forEach((tag) => {
      client.hSet(tag, key, JSON.stringify({
        lastModified: Date.now()
      }));
    });

    /**
     * Used ot set the expiration of a key based on 
     * the revalidate value or DEFAULT_TTL
     */
    client.set(
      key, 
      JSON.stringify({
        value: data,
        lastModified: Date.now(),
        tags: ctx.tags,
      }), 
      {
        EX: (ctx?.revalidate || DEFAULT_TTL),
      }
    )
  }
 
  async revalidateTag(tag) {
    const normalizedtag = tag.replace(NEXT_PREFIX, '');

    // We dont' store subkeys by page so skip check
    if (!normalizedtag.startsWith('/')) {
      // Check if a hash key exists with given tag (should only exist for FETCH)
      const keysToRemove = await client.hKeys(normalizedtag) || [];

      keysToRemove.forEach((key) => {
        // Removes the individual cache keys async if exists
        client.unlink(key);
      });
    }
  
    // Clean up hash
    client.unlink(normalizedtag);
  }
}
```

**Comment by better-salmon** [comment]
@sandervspl, @pawelblaszczyk5, @nahtnam, @duanv, @joshudev, @franco14lorenzo, @francolorenzocarestino, and @yurgenschembri-ascent, I apologize for the delay in implementation. PR #460 is now available for review and feedback.
